### DISCUSSION TITLE
To replace smooth scroll with Lenis

### DISCUSSION THREAD

**Discussion by clementroche** [AntoineW/luge#28] [Ideas]
Hi, firstly good job for the library, i think it's very useful and will help a lot of developers.
I noticed you used a CSS transform based smooth scroller, however this has negative impacts on performances, accessibility and native APIs (CSS sticky, IntersectionObserver, etc.).
At [Studio Freight](https://github.com/studio-freight) we designed a smooth scroll library called [Lenis](https://github.com/studio-freight/lenis) based on window.scrollTo approach, it provides full accessibility, has no performances cost and keeps every native APIs.
Would you like us to help you to convert the codebase to lenis in order to improve the whole library quality ?

**Comment by AntoineW** [comment]
Hi, thanks for your offer!
Your library looks great, though until now I tried to stay away from virtual-scroll, as I always thought that replacing native scroll with a virtual could lead to big troubles. That's why I chose the scroll event based way of doing it, but it has some limitations too. 

I'd like to hear your thoughts about the two techniques and the cons of virtual-scroll regarding cross-devices compatibility?

**Reply by clementroche** [reply]
We're bothered the same as you by using virtual-scroll but it's currently necessary.

But from our perspective virtual-scroll is pretty seamless for the user, Lenis is already used in production on [a bunch of websites](https://github.com/studio-freight/lenis#lenis-in-use) and we didn't receive any complains about the scroll feeling. In addition, we tested it personaly on popular environements (Windows/MacOS + mouse, trackpad and Apple Magic Mouse + Safari, Chrome, Firefox), the scroll experience resulted to be the same by adding [some extra settings](https://github.com/studio-freight/lenis/blob/main/src/lenis.js#L27-L35). I would say virtual-scroll is 100% safe, it offers almost the same scroll feeling as native and it keeps native browser features).

So of course with virtual-scroll you're not using the native scroll input but it's not very important since it feels almost the same, while knowing you'll recover accessibility, performances and native APIs i would say it's a benefic trade.

The only critical point of Lenis and smooth-scroll in general are touch devices, we advice to not use it on touch devices.

**Reply by AntoineW** [reply]
Yes I get your point. virtual-scroll is also widely used through Locomotive Scroll, so you're probably right, it's safe enough to be used now.

So, how would you suggest to integrate it in luge?

**Reply by clementroche** [reply]
I was thinking about using new Lenis(...) in [SmoothScroll.js](https://github.com/AntoineW/luge/blob/master/src/js/luge/plugins/SmoothScroll.js) and then use `lenis.on('scroll')` to trigger your own `Emitter.emit('scroll')`. So basically just use SmoothScroller as a proxy of Lenis. If I refer to [smooth scroll documentation](https://luge.cool/docs/smooth-scroll/) we just need to bind the properties to lenis properties, they share pretty much the same concepts but with a different name. For example `window.maxScrollTop` is `lenis.limit` but `window.scrollTop` remains the same because Lenis use the real window scroll value too, this should not be so difficult.

Do you have a dev environnement already setup ?

**Reply by AntoineW** [reply]
I assume it will require more than that because the concepts are slightly different. Luge looks for containers with the data-lg-smooth attribute, but if we switch to Lenis I assume that it's no longer needed as it will directly affect the global window. Basically everything has to be removed and just the `tick()` method adapted 🤔

Regarding the dev environment, I have one locally which is the same as what is on this repo. I also have some kind of demo/test site locally, but nothing clean enough to be shared as is. What would you need?

**Reply by clementroche** [reply]
About the scroll container, Lenis offers the choice too, by default it uses html but any element can be used.

I would like the luge.cool source code because i noticed all effects are used there so i can be sure everything works
